core/src/sparsity.jl

```diff
@@ -11,124 +11,34 @@ from the naming convention of this sparsity structure in the
 differentialequations.jl docs.
 """
 function get_jac_prototype(p::Parameters)::SparseMatrixCSC{Float64, Int64}
-    (; basin, pid_control) = p
+    (; basin, pid_control, graph) = p
 
     n_basins = length(basin.node_id)
     n_states = n_basins + length(pid_control.node_id)
     jac_prototype = spzeros(n_states, n_states)
 
-    for nodefield in nodefields(p)
-        update_jac_prototype!(jac_prototype, p, getfield(p, nodefield))
-    end
-
+    update_jac_prototype!(jac_prototype, basin, graph)
+    update_jac_prototype!(jac_prototype, pid_control, basin, graph)
     return jac_prototype
 end
 
 """
-If both the unique node upstream and the unique node downstream of these
-nodes are basins, then these directly depend on eachother and affect the Jacobian 2x
-Basins always depend on themselves.
-"""
-function update_jac_prototype!(
-    jac_prototype::SparseMatrixCSC{Float64, Int64},
-    p::Parameters,
-    node::Union{LinearResistance, ManningResistance},
-)::Nothing
-    (; basin, graph) = p
-
-    for id in node.node_id
-        id_in = inflow_id(graph, id)
-        id_out = outflow_id(graph, id)
-
-        has_index_in, idx_in = id_index(basin.node_id, id_in)
-        has_index_out, idx_out = id_index(basin.node_id, id_out)
-
-        if has_index_in
-            jac_prototype[idx_in, idx_in] = 1.0
-        end
-
-        if has_index_out
-            jac_prototype[idx_out, idx_out] = 1.0
-        end
-
-        if has_index_in && has_index_out
-            jac_prototype[idx_in, idx_out] = 1.0
-            jac_prototype[idx_out, idx_in] = 1.0
-        end
-    end
-    return nothing
-end
-
-"""
-Method for nodes that do not contribute to the Jacobian
-"""
-function update_jac_prototype!(
-    jac_prototype::SparseMatrixCSC{Float64, Int64},
-    p::Parameters,
-    node::AbstractParameterNode,
-)::Nothing
-    node_type = nameof(typeof(node))
-
-    if !isa(
-        node,
-        Union{
-            Basin,
-            DiscreteControl,
-            FlowBoundary,
-            FractionalFlow,
-            LevelBoundary,
-            Terminal,
-        },
-    )
-        error(
-            "It is not specified how nodes of type $node_type contribute to the Jacobian prototype.",
-        )
-    end
-    return nothing
-end
-
-"""
-If both the unique node upstream and the nodes down stream (or one node further
-if a fractional flow is in between) are basins, then the downstream basin depends
-on the upstream basin(s) and affect the Jacobian as many times as there are downstream basins
-Upstream basins always depend on themselves.
+Add nonzeros for basins connected to eachother via 1 node and possibly a fractional flow node
```

Comment:
This function is also responsible for putting nonzeros on the diagonal, basins depending on themselves, right? If so, good to make that explicit in the docstring.